### PR TITLE
Improved Client App Error Handling

### LINKED ISSUE #509: Handle API err gracefully in client app
If the API server returns an HTTP response with an error code or if the API server is unavailable the client application should handle the error gracefully and display an error message to the user.

### LINKED ISSUE #511: Angular app breaks on bad route if Elasticsearch is down
**To Reproduce**
0. Stop Elasticsearch, but keep the API up.
1. Enter a bad route into the address bar, like `local.portal.dev/badroute`
2. The page should now perpetually attempt to reload. This is because the router has no error handling for a bad route.

**Desired Behavior**
The app should gracefully display an error message to the user informing them this page does not exist, or default to the homepage. 

There is a `$stateChangeError` event so this should be a straightforward fix.

**Resources/Possible Implementations for Fixes**
http://www.mentalspike.com/2014/09/05/bad-routes-in-angular-ui-router/
http://stackoverflow.com/questions/22936865/handling-error-in-ui-routers-resolve-function-aka-statechangeerror-passing-d

### PR DESCRIPTION
**Pull request by adamcee**
Closes #511 and #509 

@joshuago78 Would you like unit tests for this? 

The app is now listening for `$stateChangeError` events which the router seems to throw in many cases, such as if a promise (like search results!) inside the router's resolve fails.
- Fails gracefully if client requests a nonexistent record
- Fails gracefully to error page if Elasticsearch is unavailable
- Fails gracefully to error page if Elasticsearch chokes/barfs on a query for some reason
- Created Error page which we can populate with messages based on the value of `error.type`

----------
PR Review

Please stress-test the app and try to break the app and see if it fails gracefully in all cases.
- [ ] @jfranco-gri 
- [ ] @joshuago78 
- [ ] @kristenGRI Kristen please let me know if you would like the look/feel of the Error Page changed.

### CODE REVIEW

File: src/client/app/app.module.js [hunk 1]
```diff
@@ -23,13 +24,26 @@
   // make lodash injectable
   .constant('_', window._)
 
-  .run(['$rootScope', '$state', '$stateParams', function($rootScope, $state, $stateParams){
+  // app initialization
+  .run(runBlock);
+
+  runBlock.$inject = ['$rootScope', '$state', '$stateParams'];
+  function runBlock($rootScope, $state, $stateParams){
```
Comment (adamcee) [inline src/client/app/app.module.js]: I could not get it to work but we should be able to put this function inside an Angular Factory and inject it into the run block. Then we'll be able to create unit tests to run against that factory and test our app initialization code.

**Comment by jfranco-gri** [comment]
This doesn't catch an error for the third scenario you have listed--if Elasticsearch chokes/barfs on a bad query. I simulated this scenario by doing these things:

1) go to src/server/api/es_functions.py
2) in create_query_string, remove one of the characters in the escape_table dictionary mapping.
3) Run a search in Portal with a query that contains the character that you removed
Expected: You go to the error page
Actual: Nothing happens, you stay at the home page. If you look at the console, you'll see errors from DataService.

Although it is unlikely that a user will be able to complete the above steps, it is possible that someone might send a bad query to Elasticsearch that we haven't covered yet. In this case, Elasticsearch doesn't throw an exception but responds with a json object that contains the error message. Django returns it as an OK response. I think we need to do something to ensure the data is good or maybe just wrap the below DataService promise code in the success function in a try block:

```
searchPromise.success(function(data) {
        var parsedData = data[0];
        parsedData.aggregations = data[1].aggregations;
        var results = SearchService.setResultsData(parsedData);
        deferred.resolve(results);
      })
```

File: src/client/app/app.module.js [hunk 2]
```diff
@@ -23,13 +24,26 @@
...
     // assign this here to persist open tabs across stage changes
     $rootScope.$activeTabs = [];
-  }]);
+
+    $rootScope.$on('$stateChangeError', function(event, toState, toParams, fromState, fromParams, error) {
+      var errorObj = {
+        error: 'NONEXISTENT_STATE',
+        data: {}
+      };
+
```
Comment (jfranco-gri) [inline src/client/app/app.module.js]: It seems like any errors in the resolves (and therefore any errors that occur when calling the api) will create the 'stateChangeError' event and will always be labeled as 'NONEXISTENT_STATE'. For instance, if the django server isn't working, but you go to `local.portal.dev:8000/search` this runs and the error is treated as a nonexistent state error, which would be misleading. 

I think for now we should just go to the error page and not create the error object. Or if we want to set up the pieces in place for the future, set the error type to 'DEFAULT' until we get more granular with our error checking and actually look at what kind of error it is.

**Comment by adamcee** [comment]
@jfranco-gri Changes made as we discussed. Tests should pass.

**Comment by joshuago78** [comment]
When I test this I don't see anything new.
- I turn off elasticsearch and try to run a query from the home page. nothing happens, no error page
- I try to look at an item detail page that does not exist and I just get the header and footer, no error page.

**Comment by adamcee** [comment]
@joshuago78 Will take a look.

**Comment by jfranco-gri** [comment]
I pulled down this branch and saw the error page when going to an item detail page that doesn't exist. I also commented out the elasticsearch import from the django api views, and that triggered the error page as well. (though there are still conflicts that need to be resolved)

**Pull request closed without merging**